### ISSUE TITLE
Nothing installed in Lineage /system/addon.d/ via Magisk app Direct Install + fix_env

### ISSUE THREAD

**Issue by mdoggydog** [topjohnwu/Magisk#3820]
I installed Magisk by using the "patch boot.img and flash with fastboot" method, since the installation instructions say that using the installer zip via a custom recovery is no longer recommended.  On the next LineageOS OTA update, Magisk was not preserved, and I had to manually patch the boot.img and flash with fastboot again.

I discovered that `/system/addon.d/99-magisk.sh` was never installed, thus the OTA update would have no way to preserve Magisk.

Digging through the code, it appears that `/system/addon.d/99-magisk.h` is only installed when a Magisk installer zip is used via a custom recovery.

A related problem:  even if Magisk is initially installed via installer-zip/custom-recovery, it appears that updating Magisk via the MagiskManager app will never update an existing `/system/addon.d/99-magisk.sh` either.

This issue is basically a duplicate of #3782, which I think got prematurely closed because I did not explain the problem well enough the first time. Please see my last comment in #3782 for more analysis of this problem.

**Comment by osm0sis** [comment]
It's supposed to install at some point via the app Direct Install too: https://github.com/topjohnwu/Magisk/blob/master/app/src/main/java/com/topjohnwu/magisk/core/tasks/MagiskInstaller.kt#L122

There was a plan to remove the unneeded installer code duplication and have everything run and maintained in one place via the shell scripts, but I'm not sure where that's at now.

**Comment by mdoggydog** [comment]
Yep... all the bits are there.  This probably needs just a one- or two-line fix `MagiskInstaller.kt`, along the lines of `cp /data/adb/addon.d.sh /system/addon.d/99-magisk.sh` spelled the correct way (well, after testing if the `/system/addon.d/` directory exists).  But maybe there is some tricky subtlety involving security contexts or file ownership or something; this android system-level stuff is all new to me.

Half of the lines of code in https://github.com/topjohnwu/Magisk/blob/b6643b7bfcb05b81fc10dc786bc7ad073e581d69/scripts/flash_script.sh#L78-L85 are just ensuring that the target directory exists and is writable; I don't know if the writeability is an issue outside of the recovery environment.

**Comment by osm0sis** [comment]
So looks like this was intentional. John said "I do not want to involve system rw at anytime during boot. This just opens up a whole can of worms..."

I won't close this as #wontfix since he may still change his mind. It would just need to use Magisk's own system mirror and rw it if the addon.d directory exists. 🤞

**Comment by mdoggydog** [comment]
I'm sorry if I'm missing something, but I just don't understand that comment in the context of this bug report, which is about install-time not boot-time.  The essential core of this issue is:
*  If your device's OS has a `/system/addon.d/` directory, then:
    * If Magisk is installed via installer .zip in recovery, then `99-magisk.sh` **is** installed.
    * If Magisk is installed via Magisk app (MagiskInstaller.kt), then `99-magisk.sh` **is not** installed.
 
That means that some devices out there are getting `99-magisk.sh` and some are not, and it only depends on how the user decided to install Magisk (and in neither case does this installation happen at boot time).  

How does "system rw at anytime during boot" come into this?

**Comment by mdoggydog** [comment]
(To rephrase again:  logically, either `99-magisk.sh` *should* be installed, or it *shouldn't*, so either MagiskInstaller.kt has a bug, or the .zip installer has a bug.)

**Comment by osm0sis** [comment]
He means "while booted", hopefully that clears up the context for you.

**Comment by mdoggydog** [comment]
Ohhhhhhh!  Yes, that clears it up a lot!  Thank you!  I now understand why it is the way it is right now.

It would be really nice if this difference between the two installation methods was mentioned in the installation documentation (https://topjohnwu.github.io/Magisk/install.html), because it is a pretty significant difference in behavior for a ROM that supports `addon.d`.  One way, Magisk magically survives future OTA updates; the other way, the user has to manually intervene before the next reboot....  

(In any case, at least the next confused schmuck like me that is trying to figure out what is going on will have this bug report to stumble upon. `:^)`)

**Comment by Tremolo4** [comment]
Could this issue appear as well when updating magisk through the app, after having initially installed via zip?

And is the only result that Magisk is not installed after the update?

Context: Me and other users of my device (`beryllium`) recently started getting bootloops after each system update. Reflashing the same system update (1), opengapps (2) and Magisk zip (3) fixes it, I'm trying to rule out Magisk.

I'm thinking maybe we were getting the bootloops because of an outdated `99-magisk.sh`, since it was never updated during magisk update via magiskmanager.

**Comment by twu2** [comment]
please re-open issue #3908 , I think it's different issue with this.

**Comment by outloudvi** [comment]
Hi. Just reporting that the survival script worked through manual LineageOS zip flashing after a fresh install of Magisk 22.
Previously since several weeks (I think maybe 5-6) ago the survival script had stopped working on my device, even after uninstalling and re-flashing Magisk's latest 21. LineageOS version is 17.1.

**Comment by ElCoyote27** [comment]
I think I'm also seeing this on LOS:
https://github.com/topjohnwu/Magisk/issues/4394
Magisk doesn't install on LOS 18.1 and 17.1 unless an old zip is used.

**Comment by brianjmurrell** [comment]
> Digging through the code, it appears that /system/addon.d/99-magisk.h is only installed when a Magisk installer zip is used via a custom recovery.

But I did install Magisk by flashing the ZIP (even though it's reported as not being supported, however [the boot patching is just too much of a palaver IMHO](https://forum.xda-developers.com/t/patch-from-rom-zip.4259695/) -- #4172) from TWRP on my Oneplus 6 (A/B device) running LineageOS 18.1 and Magisk did not survive the OTA update.

Given that I have an A/B device and OTA updating LineageOS on that also updates the recovery to use the LineageOS recovery, can the Magisk apk->zip be flashed with Lineage recovery?  I'm not at all familiar with Lineage recovery and have always used TWRP.

**Comment by osm0sis** [comment]
addon.d-v2 is only guaranteed to work from a booted OTA. Lineage didn't want it working in recovery and TWRP only has it working in the _9 branch currently.

**Comment by brianjmurrell** [comment]
@osm0sis That's what I did the last time.  Almost.  :-)  However I used TWRP recovery instead of Lineage Recovery.  Are you saying that if I flash with Lineage Recovery (i.e. `adb sideload Magisk-v23.0.zip) it will survive future OTA updates?

**Comment by brianjmurrell** [comment]
So I did install via Lineage Recovery and `adb sideload Magisk-v23.0.zip` and while flashing it did do all sorts of stuff and said it was successful at the end but looking in the Magisk app on the device it says "Installed N/A" but also says "A/B No" but we know that's not true for a Oneplus 6.

In the Magisk app though, Check SafetyNet does seem to pass but Google Pay is still saying my device is not supported even after a force stop, clear cache, clear storage.  Google Pay was working the last time I flashed with TWRP on LineageOS 18.1.

**Comment by ElCoyote27** [comment]
My recent (a few hours) experience is that 22.x and 23.x do not survive an OTA update of LOS.
In order to do that, I had to wipe and downgrade to 21.4 without updating Magisk after that.

**Comment by ossilator** [comment]
on my klte (non-a/b), v23 (installed via TWRP) survives LOS 18.1 OTA updates.

**Comment by osm0sis** [comment]
Not sure why regular addon.d (i.e. A-only devices) wouldn't work in LOS Recovery, but, again, that's a different issue from this

**Comment by Rihcus** [comment]
> @osm0sis I'm afraid I'm not up-to-speed enough to know what all of that means.
> 
> What are my options for installing Magisk on my LineageOS 18.1 flashed Oneplus6 (A/B device)? Only TWRP and [Patching Images](https://topjohnwu.github.io/Magisk/install.html#patching-images)?
> 
> Will either survive a LineageOS Updater OTA update?

Normally no but you have 3 options (ranked from easiest to hardest):

1) open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)

2) Reflash the magisk apk (rename to APK to zip) via lineage recovery (same process of flashing gapps) or flash with TWRP

3. Download and extract the boot.img from lineage os repatch with magisk manager, enter fast boot and flash (this is the most time consuming method and I personally use it as a last resort for updates).

All three methods will retain magisk preferences and modules. Hope this helps (tested lineage os 18.1 OnePlus 8)

Additional recourses:

https://topjohnwu.github.io/Magisk/install.html

**Comment by brianjmurrell** [comment]
> Normally no

That's a pity.

> 1. open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)

What do you mean exactly by "magisk should be present along with your modules and settings"?  Present where?

**Comment by 2shrestha22** [comment]
Facing same issue, I need to repeatedly pull boot.img patch and flash after Lineage OS OTA each week. Isn't anything possible to do like previous methods where we don't need to re-path boot image after each update?

**Comment by 2shrestha22** [comment]
> open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)
@Rihcus I didn't see any options like `Install after OTA`

**Comment by Rihcus** [comment]
> > open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)
> 
> @Rihcus I didn't see any options like `Install after OTA` in recent Magisk Manager version.

https://i.imgur.com/FPznfOU.png

https://i.imgur.com/wj2RHvd.jpg

Are you using the latest build of magisk manager and magisk? Also does your device support A/B?

**Comment by Rihcus** [comment]
> > Normally no
> 
> That's a pity.
> 
> > 1. open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)
> 
> What do you mean exactly by "magisk should be present along with your modules and settings"? Present where?

Basically all your settings in magisk will remain intact as will any other add-ons (in magisk manager and magisk hide).

In other words you won't have to set up magisk every time from scratch when updating with install after OTA (sorry if not clear)

**Comment by brianjmurrell** [comment]
> > What do you mean exactly by "magisk should be present along with your modules and settings"? Present where?
> 
> Basically all your settings in magisk will remain intact as will any other add-ons (in magisk manager and magisk hide).

Understood.  But that doesn't answer the actual question asked, which I left quoted above.  Does "present" mean the manager app is installed, or does it mean the .apk is available somewhere?  Is it left .apk or should it be renamed .zip?
 
> In other words you won't have to set up magisk every time from scratch when updating with install after OTA (sorry if not clear)

Again, also understood.  What is not understood is what the phrasing "magisk should be present along with your modules and settings"?  Present where and in what form?

**Comment by Rihcus** [comment]
> > > What do you mean exactly by "magisk should be present along with your modules and settings"? Present where?
> > 
> > 
> > Basically all your settings in magisk will remain intact as will any other add-ons (in magisk manager and magisk hide).
> 
> Understood. But that doesn't answer the actual question asked, which I left quoted above. Does "present" mean the manager app is installed, or does it mean the .apk is available somewhere? Is it left .apk or should it be renamed .zip?
> 
> > In other words you won't have to set up magisk every time from scratch when updating with install after OTA (sorry if not clear)
> 
> Again, also understood. What is not understood is what the phrasing "magisk should be present along with your modules and settings"? Present where and in what form?

By magisk I meant both the magisk manager app and the modified boot.img (the part which magisk manager uses to allow root access to other apps and magisk modules).

**Comment by mdoggydog** [comment]
Perhaps it would be a good idea to open a new bug report for whatever this new bug is, and move this discussion there?

This issue here is pretty specific ("nothing installed in /system/addon.d/ via..."), and is still open to track that specific issue (which is that the addon.d scripts --- whether or not they work --- are only installed in specific installation workflows).

**Comment by brianjmurrell** [comment]
> Perhaps it would be a good idea to open a new bug report for whatever this new bug is, and move this discussion there?

If it's a different problem, sure?  But is it?  I don't honestly know.
 
> This issue here is pretty specific ("nothing installed in /system/addon.d/ via..."), and is still open to track that specific issue (which is that the addon.d scripts --- whether or not they work --- are only installed in specific installation workflows).

But how do we know my issue is not in fact being caused by that also?

**Comment by 2shrestha22** [comment]
> > > open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)
> > 
> > @Rihcus I didn't see any options like `Install after OTA` in recent Magisk Manager version.
> 
> https://i.imgur.com/FPznfOU.png
> 
> https://i.imgur.com/wj2RHvd.jpg
> 
> Are you using the latest build of magisk manager and magisk? Also does your device support A/B?
> 
> 
> 
> 

I am using latest Magisk Manager Build and my device doesn't support A/B.

**Comment by osm0sis** [comment]
You guys need to stop using this as a help thread. Go use xda for troubleshooting. I'll lock this if you're not contributing anything to resolve this specific issue going forward.

**Comment by Iey4iej3** [comment]
> Hmm I believe either recovery should work. It can still install the addon.d script to /system in Lineage Recovery last time I checked.
> 
> But the Magisk apk _must_ be flashed in recovery for the addon.d script to be installed. If it's not working after that then there's something else weird going on, separate from this issue.

I don't know whether the following is acceptable: producing a flashable `zip` file which includes the patched `boot` image and the necessary `addon.d` script, and let the user flash it manually in the recovery?

**Comment by christophehenry** [comment]
So… Wait… If zip installation — which installs the `addon.d` script — is deprecated and `boot.img` flash with fastboot doesn't install the `addon.d` script, what is the correct way to have Magisk survive the updates? I don't think it is explained in the documentation. Are we supposed to reflash Magisk with `adb` after each update? This feels like a regression, no?

**Comment by 2shrestha22** [comment]
> So… Wait… If zip installation — which installs the `addon.d` script — is deprecated and `boot.img` flash with fastboot doesn't install the `addon.d` script, what is the correct way to have Magisk survive the updates? I don't think it is explained in the documentation. Are we supposed to reflash Magisk with `adb` after each update? This feels like a regression, no? 

It is explained in documentation that for non A/B devices user must reinstall Magisk. https://topjohnwu.github.io/Magisk/ota.html

**Comment by qouoq** [comment]
Alternatively one could copy [scripts/addon.d.sh](https://github.com/topjohnwu/Magisk/blob/master/scripts/addon.d.sh) to `/system/addon.d/99-magisk.sh` after having flashed a patched image. I did that two months ago and my two following updates (through TWRP) have preserved Magisk.

Needed to make /system writeable, which I did with Termux and `mount --remount / -w`, but maybe that would be safer to copy the file in TWRP.

**Comment by osm0sis** [comment]
confirmed = current behavior isn't as expected, and the recovery zip still works perfectly though isn't the recommended install method anymore.

**Comment by SPYazdani** [comment]
Hi $contact_first_name $contact_last_name,
					    We've received your case $acase_name (# $acase_case_number) on $acase_date_entered
					    Status$acase_statusReference$acase_case_numberDescription$acase_description
On 05/09/2022, 03:58, ***@***.*** wrote:confirmed = current behavior isn't as expected, and the recovery zip still works perfectly though isn't the recommended install method anymore.

**Comment by qouoq** [comment]
Would it make sense to update the install documentation to mention either or both of these methods (zip install / manual copy of addon.d script) for non A/B devices ?

**Comment by SPYazdani** [comment]
Hi $contact_first_name $contact_last_name,
					    We've received your case $acase_name (# $acase_case_number) on $acase_date_entered
					    Status$acase_statusReference$acase_case_numberDescription$acase_description
On 06/09/2022, 03:17, ***@***.*** wrote:Would it make sense to update the install documentation to mention either or both of these methods (zip install / manual copy of addon.d script) for non A/B devices ?

**Comment by osm0sis** [comment]
It still works on A/B as well.

**Comment by SPYazdani** [comment]
Hi $contact_first_name $contact_last_name,
					    We've received your case $acase_name (# $acase_case_number) on $acase_date_entered
					    Status$acase_statusReference$acase_case_numberDescription$acase_description
On 06/09/2022, 05:10, ***@***.*** wrote:It still works on A/B as well.

**Comment by infinity0** [comment]
My own `/system/addon.d/99-magisk.sh` exists but is empty. When installing via recovery, writing to `/system` using `cp` fails silently because the disk is full, even if you make it writable and use `blockdev --setrw`. (You can see the error via `adb push` however, again in recovery.) In other words, the current technique is also busted even when it is active.

I'm using LineageOS 19.1 (Android 12) with MicroG on a OnePlus 8T (KB2003), it is an A/B device. Issue originally uncovered at arovlad/bromite-webview-overlay#5.

OTA A/B install previously worked on LineageOS 18.1 (Android 11) but I never explicitly checked if `/system/addon.d/99-magisk.sh` was empty or not; I assume it wasn't and something new in Android 12 is making this fail.

**Comment by infinity0** [comment]
OTOH, the file is fine on a OnePlus 5T running LineageOS 19.1 (Android 12) without MicroG with MindTheGapps, so maybe it is a MicroG issue or a device-specific issue, I will investigate further.

Still, Magisk may want to detect that `cp` has failed (e.g. by checking the file isn't size 0) and warn the user, as it seems the busybox implementation of `cp` is a bit broken in this regard. The core root functionality works though.

**Comment by infinity0** [comment]
LineageOS devs tell me the reason is because MicroG sets WITH_GMS = true leading to lack of extra space in the system partition, so that makes sense, I'll follow this up with them. As I said though Magisk might still want to explicitly detect the failure.

**Comment by vvb2060** [comment]
This should be expected behavior, i.e. addon.d will only be written if installed via recovery. However, I think writing to the system partition is a bad idea, even in recovery.

**Comment by vvb2060** [comment]
addon.d script is part of the recovery installation. When you patch boot image, it has been indicated that you don't need addon.d script because you don't use custom recovery.

**Comment by osm0sis** [comment]
I disagree. If /system/addon.d is detected during installation of any type then there should be feature parity.

**Comment by yujincheng08** [comment]
A writable system is never the goal of Magisk.

For feature parity, I am going to remove support for addon.d on custom recovery.

**Comment by osm0sis** [comment]
That's a pretty fucked up thing to say.

**Comment by vvb2060** [comment]
> especially considering the installation instructions list recovery installs as deprecated

addon.d script is part of the recovery installation. Yes, addon.d script is also deprecated.

> so writing to system should not be concerning.

We've had issue with not having enough space to write.

addon.d script have various compatibility issues. 
And sepolicy compromises: https://github.com/topjohnwu/Magisk/blob/4d876f01458e38790eab315a32444f2efece1571/native/src/sepolicy/rules.cpp#L190-L191

**Comment by osm0sis** [comment]
Lineage runs update_engine permissive. We merely brought that to other ROMs to ensure it runs correctly.

**Comment by vvb2060** [comment]
Magisk has given up hiding, so I can point out an elephant in the room: checking /system/addon.d confirms this is a rooted device.
This is the reason why I want to delete this dir.

**Comment by yujincheng08** [comment]
> It works. Leave it alone.

Then don't expect feature parity. Magisk is never gonna remount the system as rw while booted.

FMPOV, addon.d is not even a standard (tho some custom ROMs embrace it), and it's quite *ugly* (telling all apps this is a customized ROM, disallowing using ro filesystem for the system for better compression, etc.).

I still expect to remove native support for addon.d. A third-party Magisk module should be enough for this feature.

**Comment by osm0sis** [comment]
It reveals it's a custom ROM, we're only talking about custom ROMs here. Leave attempting to hide that to the custom hiding module developers.

**Comment by yujincheng08** [comment]
> If Omni and Lineage both have a feature then it's a standard.

No.

> Leave attempting to hide that to the custom hiding module developers.

I don't think we were talking about hiding. My point is that addon.d is bad. I don't think Magisk should natively support such a bad and non-standardized workaround.

I still cannot see any reason why not leave this support to a third-party module.

**Comment by vvb2060** [comment]
> If Omni and Lineage both have a feature then it's a standard. They're the base for most of the custom ROM community.

This feature is not standard and even differs between Omni and Lineage, only Lineage's update_engine is permissive domain.

> It reveals it's a custom ROM, we're only talking about custom ROMs here, which already allow system rw without issue.

Only a small number of users are using custom ROMs. In order to be compatible with this small number of users, the sepolicy of stock ROM has been affected.

> Leave attempting to hide that to the custom hiding module developers.

This is technically impossible, at least we (shamiko developers) cannot.

**Comment by osm0sis** [comment]
Cool well a lot of custom ROM developers have put work into addon.d over the years, so it's unfortunate that you're calling their work ugly and bad.

**Comment by vvb2060** [comment]
Have any of the custom ROMs ever run CTS([Compatibility Test Suite](https://source.android.com/docs/compatibility/cts))? Do they guarantee they meet the CDD([Android Compatibility Definition Document](https://source.android.com/docs/compatibility/cdd) ) requirements?

If not, they can't even be called Android.

**Comment by osm0sis** [comment]
And there it is. Let's let Lineage know you brazenly don't consider their ROM Android.

**Comment by yujincheng08** [comment]
they are bad and ugly no matter how many people were on it. those people have no choice but to adopt this implementation. note that you hv to stay critical to customized ROMs, even for OEM ROMs. i admit the starting point of addon.d. it was good, and the implementation may be good in the past. but it cannot be good all the time.

addon.d is indeed bad and ugly. at least from for now. it deserves better implementations. this fact wont change no matter how many people have worked on it.

moreover, the whole argument was meaningless. you keep arguing about whether addon.d is good or not. but my point is whether magisk should natively support it or leave it to third-party modules. I rasie the point that addon.d is bad because i am positive to the custom ROM community that as long as Magisk hands out addon.d to third party they can come up with better implementations together with corresponding modules.

**Comment by osm0sis** [comment]
All GApps packages rely on addon.d. It's not going anywhere.

**Comment by yujincheng08** [comment]
> Removing it from Magisk just makes life on a custom ROM harder

A magisk module can relieve them, so why not.

**Comment by osm0sis** [comment]
A module that doesn't exist.

**Comment by Howard20181** [comment]
Since Android started using full disk encryption or file-based encryption, `/data/adb` is inaccessible in many cases (Magisk's OTA survival strongly depends on the scripts and files in this location) in Recovery (even LineageOS Recovery does not mount `/data`), so this feature has been unavailable for a long time.

**Comment by yujincheng08** [comment]
> A module that doesn't exist.

I will exist as soon as magisk remove addon.d support.

**Comment by osm0sis** [comment]
Hypothetically.

**Comment by yujincheng08** [comment]
> Hypothetically.

what if it comes true? Will you support removing addon.d support?

**Comment by osm0sis** [comment]
Sure, fuck it, Merry Christmas.

**Comment by osm0sis** [comment]
That said, Magisk still makes system rw on custom ROMs to remove any existing system su, so unless you have a way to systemlessly do that as part of injecting Magisk's su wherever it needs to go, then system rw is still a necessary "evil" and so the addon.d script might as well also remain part of the package.

**Comment by yujincheng08** [comment]
remove_system_su only uses on recovery mode, so not "while booted".
https://github.com/topjohnwu/Magisk/commit/17efdff134f4e1788df812dabe4d0b21f63e9c39

Further, module that add addon support for magisk: https://github.com/yujincheng08/Magisk-Addon

**Comment by osm0sis** [comment]
You're incorrect. remove_system_su is run booted as part of addon.d-v2 as well.

**Comment by yujincheng08** [comment]
> remove_system_su is run booted as part of addon.d-v2 as well

But we are now removing addon support. So when addon support removes, no more remount,rw

**Comment by osm0sis** [comment]
Then I guess you should add remove_system_su to your module.

**Comment by yujincheng08** [comment]
Since magisk has one (tho for recovery only), we can reuse it. If magisk removes this function, we can add that too. I am just proving the possibility of the existence of such a module. So the implementation is somehow trivial. Of course, we can add more features.

**Comment by osm0sis** [comment]
All of your responses have generally ranged from passive aggressive and petty, to outright open contempt throughout this debate. I'm glad @topjohnwu is still the only one who gets to press merge.

**Comment by yujincheng08** [comment]
@osm0sis You should stop personally attacking.

**Comment by osm0sis** [comment]
That's not an attack.

**Comment by yujincheng08** [comment]
Well, whether it's or not is not depending on you. But the fact is that you are the only one who says dirty words during the debate :).

**Comment by vvb2060** [comment]
@osm0sis  
> And there it is. Let's let Lineage know you brazenly don't consider their ROM Android. Maybe you should check in with topjohnwu before you alienate the entire Android customization community.

It looks like LineageOS won't pass CTS and I'm surprised by the news. LineageOS, being the most popular custom ROM, should avoid splitting the Android community and make the ROM compliant with CDD to reduce the pressure on developers to adapt it.

**Comment by vvb2060** [comment]
https://github.com/programminghoch10/Lygisk/blob/ci-management/README.md?plain=1#L9-L11

I saw the previous link programminghoch10/Lygisk#20 , this feature was always broken.

**Comment by christophehenry** [comment]
That is a pretty heated conversation here. Very technical too, but I haven't seen anyone consider the problem from the user perspective. It excessively simple: *I don't want to be required to reinstall Magisk any time I do a Lineage update*. This is a fairly legitimate request, I think. I don't care how ugly `addons.d` is or how hypothetically it may break things in the future.

It works.

And that's all a user asks for. So, until a better solution is found, there's no good reason to deny a feature parity based on that one.

**Comment by yujincheng08** [comment]
@christophehenry I absolutely know users' requests. And my point is to avoid magisk doing such a dirty job. Instead, a magisk module is enough for this job. We have basically developed a very initial version of such a module: https://github.com/yujincheng08/Magisk-Addon. Moreover, other developers are now attracted to make this module not depend on an unencrypted data partition.

**Comment by osm0sis** [comment]
My responses have been quite cordial by comparison despite my using of the occasional swear. 🤔

**Comment by christophehenry** [comment]
@yujincheng08 Sounds promising.

Thank you all for your work.

**Comment by mdoggydog** [comment]
Hi, it's me, 🤩 @mdoggydog, the original poster here at good ol' #3820 --- "the thread that keeps on giving". 🎅🎁

As a _conscientious but naive user_, my perspective is:
 * I don't follow Magisk conversations all day on XDA/Twitter/Reddit/etc, but I do my best to RTFM (read the documentation) before I go pestering anyone.
   * The Magisk README literally says 
     > "[Github](https://github.com/topjohnwu/Magisk/) is the only source where you can get official Magisk information and downloads."

     so I don't think I should even try to look anywhere else.
 * "Surviving OTA updates" is an essential feature for Magisk, and should be supported on any system where it is possible.
   * I am using a ROM that features OTA updates, and I chose that ROM in part because I want those easy OTA updates; I bothered to install Magisk, because I need Magisk; and thus, if Magisk does not survive an OTA update (especially if it _could_ and it _used to_), I will be frustrated.
 * "Via `.zip` sideload in Recovery" is the best way to install Magisk, and I have no idea why it would be deprecated.
   * Why the best?  Because it is so simple!  Look:
     ```
     $ adb reboot recovery
     $ adb sideload Magisk-vXXX.zip
      ```
      That's it!  And if I just installed a ROM via Recovery, then I only need the second line!  No craziness with extracting a boot.img from _somewhere_, transfer to phone, patch it, transfer it back, boot to bootloader, flash via fastboot... Who wants to do all that?
    * Seriously, "Install via `.zip` in Recovery" is awesome.
 * The documentation says nothing about what happens to Magisk after a system update (OTA or not), so if Magisk disappears after an update, I become worried; and if Magisk had survived earlier updates and then disappears, I am confused; and when I finally discover an answer by digging through two-year-old bug reports in github, and I realize that I am just one of many users suffering the same experience, I am frustrated and angry.
   * Yes, there is that `docs/ota.md` page, but _it is not linked by any of the documentation index pages_.  I only find it after desperately rooting around in the Magisk source tree.  (Even _I_ keep forgetting it exists, even though this is my favorite topic on my favorite piece of software of all time.)  Also, it doesn't mention anything about the addon.d mechanism.
   * To this day, the documentation still does not mention, anywhere, that "Install via `.zip` in Recovery" is the magic that provides the addon.d installation.
 * I know nothing about Modules and the Magisk documentation does not help me.  I do not know where to find modules, how to install them, or (very important) how to assess the trustworthiness of any modules I do find.

As a _security-conscious software engineer_, my perspective is:
 * Moving OTA-survival into a module seems reasonable because:
   1. As I understand it, OTA-survival will never work automagically on StockROM's, because stock Android has no native mechanism for "specify extra stuff to do at the end of an OTA update", so it cannot be a "universal" feature anyway.
   2. Implementation as a module would make it easier to support different/new/additional OTA-survival mechanisms in the future.
   3. Set up via installing a module makes it easier to have a single user-flow for enabling OTA-survival, when it is desired and available, no matter how Magisk was installed.  (I.e., "Finally, if you want OTA-survival, install one of these modules...")
 * _However,_ I am concerned that if an essential feature is moved into a module, it will become more __fragile__ because Magisk-core will take the attitude of "not my problem anymore!"
   * The basic machinery for OTA-survival needs to be part of the core, and needs to have a stable API.   (I assume that there is a lot of overlap between "bits needed to install Magisk" and "bits needed to survive OTA".)
   * The focus of survival modules should be on how/when that core machinery is activated.
   * The Magisk documentation should, at the very least, inform users that these modules exist and where to find them.
 * I would not want to see an essential feature, OTA-survival, removed from Magisk-core until there was a well-tested, widely-available, well-documented alternative available.
   * I appreciate that @yujincheng08 has a proof-of-concept module already, but I wouldn't expunge addon.d support from Magisk-core yet without the testing, documenting, and informing users.
 * "Deprecation" means (1) informing other people, through documentation, that a change will be happening; (2) providing a timeline for that change; (3) providing the alternative that should be used to replace the deprecated element.  Deprecation is a process to ensure feature-continuity without surprises.
   * Mentioning that something is deprecated in a sentence buried in 87 comments in a bug report is, at best, "wishful thinking", at worst, "preparing for failure".

Happy Holidays to all the residents of #3820! 
# 😷 🥳 🎉 🍾 🍿

**Comment by osm0sis** [comment]
Okay, so the best way to handle this would be that addon.d script becomes opt-in only as a button in the app settings, similar to Systemless Hosts. That way it can be used regardless of Magisk install method, and a custom ROM user can choose not to use it too.

**Comment by breversa** [comment]
I just found out about this issue while doing yet another Magisk reinstall after a LineageOS-based [/e/OS](https://e.foundation/e-os/) update.

Is there any roadmap to implement @osm0sis’ proposed solution?

@mdoggydog : best user [feedback](https://github.com/topjohnwu/Magisk/issues/3820#issuecomment-1366226450)/POV I’ve ever read!!! :D

**Comment by brianjmurrell** [comment]
Looking on my device in `/system/addon.d/` I don't see any `99-magisk.sh`.

So assuming I have Magisk 26.3 installed on my device, if I (i.e. manually, with `adb`) copy https://github.com/topjohnwu/Magisk/blob/v26.3/scripts/addon.d.sh to `/system/addon.d/99-magisk.sh` on my device, can I stop having to do:

open magisk manager before starting lineage os updater (important), minimize magisk manager and run lineage os updater do not reboot when prompted, go back to magisk manager click install magisk and click install after OTA and reboot when prompted (magisk should be present along with your modules and settings)

and just let the normal post-OTA-update reboot happen as usual and have Magisk survive the OTA update?  My device is A/B if that makes any difference.

**Comment by brianjmurrell** [comment]
> > Further, module that adds addon support for magisk: https://github.com/yujincheng08/Magisk-Addon.

This one supplies it's own script for `/data/addon.d`.  What happens if a newer Magisk updates it's `addon.d` script?
 
> Found a module that works around the issue: https://github.com/mModule/mSurvival

This one installs the `addon.d` script that is supplied by the Magisk that is installed.  This seems more resilient than the above one supplying it's own.  But it still begs the question of what happens if Magisk is updated to a newer version and that newer version of Magisk supplies an updated `addon.d` script.

**Comment by optimumpr** [comment]
Why can't Magisk simply patch both slots the same way TWRP installers do?

**Comment by Joni-Ahmed-Dev** [comment]
I think It's ok.

**Comment by duven87** [comment]
Hi, I noticed that some apps (tk.de) checks if magisk.sh is in addon.d and does not let you use warning that the phone is rooted.
Would it be possible that with the hide Magisk app option also edit that script name?

**Comment by tamer-hassan** [comment]
> Hi, I noticed that some apps (tk.de) checks if magisk.sh is in addon.d and does not let you use warning that the phone is rooted. Would it be possible that with the hide Magisk app option also edit that script name?

simply try:
adb root
adb remount
adb shell
# chown root:root /system/addon.d
# chmod 700 /system/addon.d

and reboot and try the TK app again
if that doesn't work, try:
# mv /system/addon.d /system/add-on.d
but remember to rename it back to original name before any OTA update.

**Comment by topjohnwu** [comment]
I'm closing this thread because I'm still taking the stance that Magisk do not want to involve system rw at anytime during boot.

**Comment by ossilator** [comment]
this justification leaves me somewhat guessing despite being a developer.
please provide an explanation (possibly in the form of a link).
please also state what you see as the best alternative, specifically vs. the excellent breakdown in https://github.com/topjohnwu/Magisk/issues/3820#issuecomment-1366226450.
